observability/akka-http-metrics/src/main/scala/com/daml/metrics/akkahttp/AkkaHttpMetrics.scala

```diff
@@ -0,0 +1,126 @@
+    Directive { (fn: Unit => Route) => ctx =>
+      // process the query, using a copy of the httpRequest, with size metric computation
+      val newCtx = ctx.withRequest(
+        ctx.request.withEntity(
+          requestEntityContentLenghtReportMetric(ctx.request.entity, requestBytesTotal)
+        )
+      )
+      val result = Timed.future(latency, fn(())(newCtx))
+
+      result.transform { result =>
+        result match {
+          case scala.util.Success(Complete(httpResponse)) =>
+            // record request
+            requestsTotal.inc()
+            if (httpResponse.status.isFailure)
+              // record failure
+              errorsTotal.inc()
+            // return a copy of the httpResponse, with size metric computation
+            scala.util.Success(
+              Complete(
+                httpResponse.withEntity(
+                  responseEntityContentLenghtReportMetric(httpResponse.entity, responseBytesTotal)
+                )
+              )
+            )
+          case _ =>
+            // record request and failure
+            requestsTotal.inc()
+            errorsTotal.inc()
+            result
+        }
+      }
+
+    }
+
+  // support for computation and report of the size of a requestEntity
+  // for streaming content, creates a copy of the requestEntity, with embedded support
+  private def requestEntityContentLenghtReportMetric(
+      requestEntity: RequestEntity,
+      metric: Counter,
+  )(implicit mc: MetricsContext): RequestEntity =
+    requestEntity match {
+      case e: HttpEntity.Default =>
+        e.copy(data = byteStringSourceLenghtReportMetric(e.data, metric))
+      case e: HttpEntity.Strict =>
+        metric.inc(e.data.length.toLong)
+        e
+      case e: HttpEntity.Chunked =>
+        e.copy(chunks = chunkStreamPartSourceLengthReportMetric(e.chunks, metric))
+
+    }
+
+  // support for computation and report of the size of a responseEntity
+  // for streaming content, creates a copy of the responseEntity, with embedded support
+  private def responseEntityContentLenghtReportMetric(
+      responseEntity: ResponseEntity,
+      metric: Counter,
+  )(implicit mc: MetricsContext): ResponseEntity =
+    responseEntity match {
+      case e: HttpEntity.Default =>
+        e.copy(data = byteStringSourceLenghtReportMetric(e.data, metric))
+      case e: HttpEntity.Strict =>
+        metric.inc(e.data.length.toLong)
+        e
+      case e: HttpEntity.Chunked =>
+        e.copy(chunks = chunkStreamPartSourceLengthReportMetric(e.chunks, metric))
+      case e: HttpEntity.CloseDelimited =>
+        e.copy(data = byteStringSourceLenghtReportMetric(e.data, metric))
+    }
+
+  // adds a side flow to the source, to compute and report the total size of the ByteString elements
+  private def byteStringSourceLenghtReportMetric[Mat](
+      source: Source[ByteString, Mat],
+      metric: Counter,
+  )(implicit mc: MetricsContext): Source[ByteString, Mat] =
+    source.alsoTo(
```

Comment:
The simple `wireTap` function doesn't work here, because a `fold` is needed to compute the aggregated size of all the element of the source.
The complete `wireTap` function has the same signature as `alsoTo`. The difference is that `wireTap` can drop elements if the additional sink is too slow. `alsoTo` links the backpressure of the additional sink to the main flow.

So `wireTap` would drop the metrics reporting if the system is overloaded. It wouldn't be a great behavior, the metrics would not be complete when they are likely needed to understand the load of the system.

Reply:
It doesn't create a new stream, it adds a fork to the existing stream.
When materialized, everything will run together in the same context.

Duplicating the streams in fact doesn't work, because some operations (like upload dar) don't support reading the sources more than once.

Reply:
Fair enough.

Not fully convinced about `alsoTo` vs `wiretap` though. 
If somehow the metrics aren't processed fast enough, do we want to take the risk of applying backpressure to the stream itself? I would much rather lose metrics, and maybe record somehow that we're losing metrics

Reply:
That sounds like a great approach, i.e., if we could make it transparent when the metrics subsystem cannot keep up, then dropping some metric samples should be acceptable. Perhaps akka already has some logging / metrics that we could leverage for this?

Reply:
In general, I am not sure that loosing metrics when the system is overloaded is a good policy.
One of the usage of the metrics is to give insight in the system when there are issues.
If the metric is not available because there are issues, then it becomes fairly useless.

Reply:
I cannot find a simple way to be informed that backpressure caused a element to be dropped.
`Flow.buffer` allows to drop elements in case of backpressure, but there is no feedback
`Source.queue` can be used to create a new flow with backpressure feedback. But it requires running an entire new stream. [documentation](https://doc.akka.io/docs/akka/current/stream/operators/Source/queue.html#source-queue)

Reply:
Would conflate be a possible option? https://doc.akka.io/docs/akka/current/stream/operators/Source-or-Flow/conflate.html

Reply:
It is not quite the purpose of `conflate`.
I think we can express the backpressure information with it. But it might make the problem worse ?
We would go from
```
fold(0L)((acc, d) => acc + d.length).to(Sink.foreach(metric.inc(_))
```
to
```
conflate{(acc, _) => backpressuremetric.inc; acc}.fold(0L)((acc, d) => acc + d.length).to(Sink.foreach(sizemetric.inc(_))
```
I think it adds an async boundary, which makes the stream more expensive. And if the fold is not fast enough (?) because the system is too busy, it adds a metric update for every backpressured element, and the fold is still being processed for the other elements.